python/metatensor-operations/metatensor/operations/block_from_array.py

```diff
@@ -73,31 +96,67 @@ def block_from_array(array) -> TensorBlock:
             # we are using metatensor-core
             labels_array_like = np.empty(0)
 
+    shape = array.shape
+    n_dimensions = len(shape)
+    if n_dimensions < 2:
+        raise ValueError(
+            f"the array provided to `block_from_array` \
+            must have at least two dimensions. Too few provided: {n_dimensions}"
+        )
+
+    # constructs the default label names and counts
+    if sample_names is None:
+        sample_names = ["sample"]
+    d_samples = len(sample_names)
+    if property_names is None:
+        property_names = ["property"]
+    d_properties = len(property_names)
+    # guess number of components
+    d_components = n_dimensions - d_samples - d_properties
+    if d_components < 0:
+        raise ValueError(
+            f"the array provided to `block_from_array` with shape {shape} "
+            + "does not have enough dimensions to match the sample and property names"
+        )
+    if component_names is None:
+        component_names = [
+            f"component_{component_index+1}" for component_index in range(d_components)
+        ]
+    if len(component_names) != d_components:
+        raise ValueError(
+            f"the array provided to `block_from_array` with shape {shape} "
+            + "does not have enough dimensions to match the given sample, "
+            + "component, and property names"
+        )
+
     samples = Labels(
-        names=["sample"],
-        values=_dispatch.int_array_like(
-            list(range(shape[0])), labels_array_like
-        ).reshape(-1, 1),
+        names=sample_names,
+        values=_dispatch.indices_like(shape[0:d_samples], labels_array_like),
     )
     components = [
         Labels(
-            names=[f"component_{component_index+1}"],
+            names=[component_names[component_index]],
             values=_dispatch.int_array_like(
                 list(range(axis_size)), labels_array_like
             ).reshape(-1, 1),
         )
-        for component_index, axis_size in enumerate(shape[1:-1])
+        for component_index, axis_size in enumerate(shape[d_samples:-d_properties])
     ]
     properties = Labels(
-        names=["property"],
-        values=_dispatch.int_array_like(
-            list(range(shape[-1])), labels_array_like
-        ).reshape(-1, 1),
+        names=property_names,
+        values=_dispatch.indices_like(shape[-d_properties:], labels_array_like),
     )
 
     device = _dispatch.get_device(array)
     samples = samples.to(device)
     components = [component.to(device) for component in components]
     properties = properties.to(device)
+    block_shape = [len(samples)]
+    for i in range(d_samples, d_samples + d_components):
+        block_shape.append(shape[i])
+    block_shape.append(len(properties))
+    # reshape array if sample of property axes need to be merged
+    if len(block_shape) != len(shape):
+        array = array.reshape(block_shape)
```

Comment:
Sorry, I missed this earlier. I would rather not reshape here. Instead the user can reshape before calling the function if they want to. If you agree, I can make the change (I found this while cleaning up the code locally)